### PR TITLE
Quant studio plate prep2

### PR DESCRIPTION
**Pull request by EGCGDanForth**
New script for autoplacement of up to 27 input plates into  1 96 well plate.
New script for generating hamilton input for up to  27 input plates into up to 3 files

### CODE REVIEW

File: EPPs/common.py [hunk 1]
```diff
@@ -385,6 +387,76 @@ def _run(self):
...
+        # create a counter so only use each available well in the output plate once (24 wells available in the 96 well plate
+        well_counter = 0
+
+        # loop through the input containers and place the samples in row-column order - this makes pipetting as efficient
+        # as possible, particularly if only 1 input plate so 1:1 pipetting
+        for container in sorted(input_container_nested_dict, key=lambda x: x.name):
+            for column in self.input_plate_layout_columns:
+                for row in self.input_plate_layout_rows:
+                    # populate list of tuples for set_placements if well exists in input plate
+                    if row + ":" + column in input_container_nested_dict[container].keys():
```
Comment (tcezard) [inline EPPs/common.py]: no need for the function call `.keys()` here

File: scripts/autoplacement_24_in_96.py [hunk 2]
```diff
@@ -1,72 +1,18 @@
 #!/usr/bin/env python
-import sys
 
-from EPPs.common import StepEPP
+from EPPs.common import Autoplacement
 
 
 # Script for performing autoplacement of samples. If 1 input and 1 output then 1:1 placement. If multiple input plates then
 # takes all samples from each plate before next plate Loops through all inputs and assigns them to the next available space
 # by column-row in the output plate
```
Comment (tcezard) [inline scripts/autoplacement_24_in_96.py]: Can you put these comments in the class docstring:
```python
class Autoplacement24in96(Autoplacement):
    """
    Script for performing autoplacement of samples. If 1 input and 1 output then 1:1 placement. If multiple input plates then
    takes all samples from each plate before next plate Loops through all inputs and assigns them to the next available space
    by column-row in the output plate
    """"

File: scripts/autoplacement_96_well.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env python
+
+from EPPs.common import Autoplacement
+
+
+# Script for performing autoplacement of samples from multiple input plates into a single output plate.
+# If 1 input and 1 output then 1:1 placement. If multiple input plates then
+# takes all samples from each plate before next plate Loops through all inputs and assigns them to the next available space
+# by column-row in the output plate
+class Autoplacement96well(Autoplacement):
```
Comment (tcezard) [inline scripts/autoplacement_96_well.py]: Same comment as above, the comment should be a docstring

File: scripts/generate_hamilton_input_quant_studio.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        assert self._max_nb_input_containers is not None, 'number of permitted input containers needs to be set ' \
+                                                          'by the child class'
+        assert self._max_nb_output_containers is not None, 'number of permitted output containers needs to be set ' \
+                                                           'by the child class'
+
+    @staticmethod
+    def add_args(argparser):
+
+        argparser.add_argument(
+            '-i', '--hamilton_input', nargs='*', help='Add up to 3 file spaces in LIMS step for Hamilton inputs')
```
Comment (tcezard) [inline scripts/generate_hamilton_input_quant_studio.py]: `hamilton_input` should be pluralised here and in the variable name to make it clear it is a list.
I think you'd  need to have at least one file so change `nargs` to `nargs='+'`

File: scripts/generate_hamilton_input_quant_studio.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    def generate_csv_dict(self):
+        csv_dict = {}
+
+        for art in self.artifacts:
+            if art.type == 'Analyte':
+                output = self.process.outputs_per_input(art.id, Analyte=True)
+                # the script is only compatible with 1 output for each input i.e. replicates are not allowed
+                if len(output) > 1:
+                    raise InvalidStepError('Multiple outputs found for an input %s. '
```
Comment (tcezard) [inline scripts/generate_hamilton_input_quant_studio.py]: This check can be pushed to `_nb_analytes_per_input=1`

File: tests/test_generate_hamilton_input_quant_studio.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
+from itertools import cycle
+
+from scripts.generate_hamilton_input_quant_studio import GenerateHamiltonInputQuantStudio
+from tests.test_common import TestEPP, FakeEntitiesMaker
+
+
+class TestGenerateHamiltonInputQuantStudio(TestEPP):
+    def setUp(self):
+        # argument -d left blank to write file to local directory
```
Comment (tcezard) [inline tests/test_generate_hamilton_input_quant_studio.py]: There is no -d argument in `GenerateHamiltonInputQuantStudio`

File: EPPs/common.py (hunk 1 above)
Comment (EGCGDanForth) [reply EPPs/common.py]: updated

File: scripts/autoplacement_24_in_96.py (hunk 2 above)
Comment (EGCGDanForth) [reply scripts/autoplacement_24_in_96.py]: updated

Comment (EGCGDanForth) [reply scripts/autoplacement_24_in_96.py]: updated

File: scripts/generate_hamilton_input_quant_studio.py (hunk 4 above)
Comment (EGCGDanForth) [reply scripts/generate_hamilton_input_quant_studio.py]: updated

File: scripts/autoplacement_96_well.py (hunk 3 above)
Comment (EGCGDanForth) [reply scripts/autoplacement_96_well.py]: updated

File: scripts/generate_hamilton_input_quant_studio.py (hunk 5 above)
Comment (EGCGDanForth) [reply scripts/generate_hamilton_input_quant_studio.py]: updated

File: tests/test_generate_hamilton_input_quant_studio.py (hunk 6 above)
Comment (EGCGDanForth) [reply tests/test_generate_hamilton_input_quant_studio.py]: updated

**Comment by EGCGDanForth** [comment]
For last commit 47681967ab1179978afef49529be395754c9415d, the comment should also have stated that that test_normalisation.py was not correctly testing that input_conc can be obtained from sample udf.

File: scripts/normalisation.py [hunk 7]
```diff
@@ -44,18 +43,29 @@ def _run(self):
...
-                output.udf[self.output_buffer] = 0
-            else:
-                output.udf[self.output_volume] = round(
-                    (target_volume * (target_concentration / art.udf.get(self.input_conc))), 1)
+            input_conc = art.udf.get(self.input_conc)
+            if not input_conc:
+                input_conc = art.samples[0].udf.get(self.input_conc)
 
-                output.udf[self.output_buffer] = target_volume - output.udf.get(self.output_volume)
+            if all(v is not None for v in [input_conc,target_concentration,target_volume]):
```
Comment (tcezard) [inline scripts/normalisation.py]:
```suggestion
            if None not in [input_conc, target_concentration, target_volume]:
```

**Pull request merged**